iris-ugrid/ugrid_cf_reader.py

```diff
@@ -0,0 +1,231 @@
+            "cube_dim",
+            "grid",
+            "mesh_location",
+            "topology_dimension",
+            "node_coordinates",
+        ],
+    )
+):
+    """
+    Object recording the unstructured grid dimension of a cube.
+    * cube_dim (int):
+        The cube dimension which maps the unstructured grid.
+        There can be only one.
+    * grid (`gridded.pyugrid.UGrid`):
+        A 'gridded' description of a UGRID mesh.
+    * mesh_location (str):
+        Which element of the mesh the cube is mapped to.
+        Can be 'face', 'edge' or 'node'.  A 'volume' is not supported.
+    * topology_dimension (int):
+        The highest dimensionality of the geometric elements in the mesh.
+    * node_coordinates (list):
+        A list of the names of the spatial coordinates, i
+        used to geolocate the nodes.
+    """
+
+    def __str__(self):
+        result = "Cube unstructured-grid dimension:"
+        result += "\n   cube dimension = {}".format(self.cube_dim)
+        result += '\n   mesh_location = "{}"'.format(self.mesh_location)
+        result += '\n   mesh "{}" :'.format(self.grid.mesh_name)
+        result += '\n   topology_dimension "{}" :'.format(
+            self.topology_dimension
+        )
+        result += '\n   node_coordinates "{}" :\n'.format(
+            " ".join(self.node_coordinates)
+        )
+        try:
+            mesh_str = str(self.grid.info)
+        except TypeError:
+            mesh_str = "<unprintable mesh>"
+        result += "\n".join(["     " + line for line in mesh_str.split("\n")])
+        result += "\n"
+        return result
+
+    def name(self):
+        return ".".join([self.grid.mesh_name, self.mesh_location])
+
+
+class UGridCFReader:
+    """
+    A CFReader extension to add UGRID information to netcdf cube loading.
+    Identifies UGRID-specific parts of a netcdf file, providing:
+    * `self.cfreader` : a CFReader object to interpret the CF data from the
+      file for cube creation, while ignoring the UGRID mesh data.
+    * `self.complete_ugrid_cube(cube)` a call to add the relevant UGRID
+      information to a cube created from the cfreader data.
+    This allows us to decouple UGRID from CF support with minimal changes to
+    the existing `iris.fileformats.netcdf` code, which is intimately coupled to
+    both the CFReader class and the netCDF4 file interface.
+    """
+
+    def __init__(self, filename, *args, **kwargs):
+        self.filename = os.path.expanduser(filename)
+        dataset = netCDF4.Dataset(self.filename, mode="r")
+        self.dataset = dataset
+        meshes = {}
+        for meshname in find_mesh_names(self.dataset):
+            mesh = UGrid()
+            load_grid_from_nc_dataset(dataset, mesh, mesh_name=meshname)
+            meshes[meshname] = mesh
+        self.meshes = meshes
+
+        # Generate list of excluded variable names.
+        exclude_vars = list(meshes.keys())
+
+        temp_xios_fix = kwargs.pop("temp_xios_fix", False)
```

Comment:
As discussed, and written up in #5, `temp_xios_false` is not planned to be used this time round. This code block should be reduced to just the standard handling (when `temp-xios-fix=False`.